src/pykeen/nn/node_piece.py

```diff
@@ -0,0 +1,445 @@
+                tuple(edge_index),
+            ),
+            shape=(num_entities, num_entities),
+        )
+        # compute distances between anchors and all nodes, shape: (num_anchors, num_entities)
+        distances = scipy.sparse.csgraph.shortest_path(
+            csgraph=adjacency,
+            directed=False,
+            return_predecessors=False,
+            unweighted=True,
+            indices=anchors,
+        )
+        # select anchor IDs with smallest distance
+        return torch.as_tensor(numpy.argpartition(distances, kth=min(k, num_entities), axis=1), dtype=torch.long)
+
+
+class ScipySparseAnchorSearcher(AnchorSearcher):
+    """Find closest anchors using scipy.sparse."""
+
+    def __init__(self, max_iter: int = 5) -> None:
+        self.max_iter = max_iter
+
+    def __call__(self, edge_index: numpy.ndarray, anchors: numpy.ndarray, k: int) -> numpy.ndarray:  # noqa: D102
+        # infer shape
+        num_entities = edge_index.max().item() + 1
+        # create adjacency matrix
+        adjacency = scipy.sparse.coo_matrix(
+            (
+                numpy.ones_like(edge_index[0], dtype=bool),
+                tuple(edge_index),
+            ),
+            shape=(num_entities, num_entities),
+        )
+        # symmetric + self-loops
+        adjacency = adjacency + adjacency.transpose() + scipy.sparse.eye(num_entities, dtype=bool, format="coo")
+        adjacency = adjacency.tocsr()
+        logger.info(f"Created adjacency matrix: {adjacency}")
+
+        # for each entity, determine anchor pool by BFS
+        num_anchors = len(anchors)
+        pool = numpy.zeros(shape=(num_entities, num_anchors), dtype=bool)
+        reachable = numpy.zeros(shape=(num_entities, num_anchors), dtype=bool)
+        reachable[anchors] = numpy.eye(num_anchors, dtype=bool)
+        final = numpy.zeros(shape=(num_entities,), dtype=bool)
+
+        for _i in range(self.max_iter):
+            # propagate one hop
+            reachable = adjacency.dot(reachable)
+            # copy pool if we have seen enough anchors and have not yet stopped
+            num_reachable = reachable.sum(axis=1)
+            enough = num_reachable >= k
+            mask = enough & ~final
+            pool[mask] = reachable[mask]
+            # stop once we have enough
+            final |= enough
+        del reachable, final
+
+        tokens = numpy.full(shape=(num_entities, k), fill_value=-1)
+        # select from pool
+        # use sparse random sampling due to memory footprint
+        entity_ids, anchor_ids = pool.nonzero()
+        unique_entity_ids, counts = numpy.unique(entity_ids, return_counts=True)
+        assert (counts >= k).all()
+        generator = numpy.random.default_rng()
+        intra_offset = numpy.floor(
+            generator.random(size=(counts.size, k), dtype=numpy.float32) * counts[:, None]
+        ).astype(int)
+        offset = numpy.cumsum(numpy.r_[0, counts])[:-1, None] + intra_offset
```

Comment:
Honestly, this whole procedure is very cryptic to me - I don't understand what's going on here with those `r_` ,`cumsum` and generators 😅 
Besides, I don't see where the anchors are actually sorted by shortest path distances, the above loop in `max_iter` iterations just keeps track of reachable anchors, not their distances

Reply:
* `Generator` is the preferred way to generate random numbers in newer numpy versions, cf., https://numpy.org/doc/stable/reference/random/generator.html
* the method does a BFS until it reaches at least `k` anchor nodes, or the maximum number of iterations is exceeded -> this is the `pool` for each entity
* then it randomly samples from the pool. Unfortunately, I could not create a dense mask and multiply that by the mask to do the selection, since this led to OOM. So this `cumsum` part, does some sparse version of that; but this cannot make sure anymore that no duplicates are selected.

Reply:
I think we should just keep the `csgraph` solution, and leave further scalability for a future PR :sweat_smile: